### DISCUSSION TITLE
Can I solve Ax=b1,b2,b3 repeatedly on GPU?

### DISCUSSION THREAD

**Discussion by ztdepztdep** [ginkgo-project/ginkgo#1009] [Q&A] [answered]
I need to solve a series of Ax=bi,i=1.....N with the matrix A unchanged. 
             `while(time<FianalTime)
               {
                    solve Ax=bi;                   
                    set bi;
                     if(time=SaveTime) copy x to cpu for further treating
              }`
I want to know can i set the bi directly on the GPU without copying data between cpu and gpu.

**Comment by upsj** [comment, marked as answer]
Yes, what you are describing is possible directly! All data that is allocated on the GPU can be accessed via the corresponding data pointers `vector->get_values()`. So depending on the complexity of the operations you want to do on `b_i`, you can write your own CUDA kernels, build the operations out of [thrust](https://github.com/NVIDIA/thrust) algorithms or (probably soon) use our [common kernel interface](https://github.com/ginkgo-project/ginkgo/pull/938) to implement simple pointwise kernels in a device-independent fashion.

Let us know if you want some details on any of the 3 approaches. What kinds of calculations do you need to do to compute the new `b_i`?

**Reply by ztdepztdep** [reply]
ok, I am solving Navier-Stokes equations, the bi is updated by computing the gradient, and divergence of the velocity vector fields and it is very suited to write a CUDA kernels by myself and transfer the array to the library.
